### PR TITLE
fix(semantic): reference flags not correctly resolved when after an export stmt

### LINKED ISSUE #7879: Regression in 0.15.1 typescript-eslint(consistent-type-imports)
A number of imports are considered type-only imports, but they are actually used as values.

See https://github.com/toeverything/AFFiNE/actions/runs/12317223993/job/34379255779?pr=9161

### PR DESCRIPTION
**Pull request by camc314**
fixes https://github.com/oxc-project/oxc/issues/7879#issuecomment-2562574889

TLDR is curently here:

https://github.com/oxc-project/oxc/blob/cdd121bfa4a66d5b2cf72a444f35e82daf81d11e/crates/oxc_semantic/src/builder.rs#L2130-L2135

`self.current_reference_flags.is_empty()` is not empty, which causes the current ref flags to be used (this is incorrect, we should be using a fresh version of reference flags). Buy setting ref flags to `None` on exit export node, this issue is avoided

`export` **BEFORE** the reference ( incorrect reference flags)

https://playground.oxc.rs/#eNpVjjuOwzAMRK8isElj7A/YxtttkVOkkR3aECCJBskkdgzdPZISG0ilGc3DcFbooQUXJmI1q/m3bJIZmII5fHx2lg9/p4hzTXWZsCL3N+RekB3qvRUxRyKDs2I8S61cEzRA0K7Al1geWaLaGVrlCzbgXdRNS08T7mYJHfnNKdsoA3GAdrBeMDUwWRbk3Jh1adn0jtYPUMsj5hOA8vP1/QuZ6OmMI5Yx2QQX3eCebLBx9K8FlYvK5I+ebiW9InckOX4uSOkBCrdvkw==

`export` **AFTER** the reference ( correct reference flags)

https://playground.oxc.rs/#eNpVjj2uwjAQhK9ibZMmen/Sa0JHwSlonLCJLNm70a6BhMh3xzEEQeUZz6fZWaCDBlwYWaJZzN6KSaYXDqb6+m6tVLsjHQmnknfeqpoDs8EpIp208Et6Q+I8Yum5ffTcqh3UwNAsIGdaH50p2gmaKGeswTuKm9aOR3yZObTsNxfFkvYsAZreesVUw2hFUXJj1mvLpl9o+YBoZcB8AlD/fn7/IRMdn3DAdUw2wZHr3YMNlgb/XFA4isL+4Pm6pheUljXHjwUp3QEtmnBd

this PR fixes this issue by resetting the reference flags after exising an export stmt

### CODE REVIEW

**Comment by camc314** [comment]
* **#8134** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8134?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8134?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8134).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by Dunqing — CHANGES_REQUESTED** [review]
Thank you for working on this!

> self.current_reference_flags.is_empty() is not empty, which causes the current ref flags to be used (this is incorrect, we should be using a fresh version of reference flags). Buy setting ref flags to None on exit export node, this issue is avoided

This is not the root cause, after resolving reference flags, the flags should be empty because it took, so the problem is why it still has `ReferenceFlags::Type`. After debugging, I found the problem is related to export if has `source`, you can see [playground](https://playground.oxc.rs/#eNo9T8tuhDAM/JXIZ9SX1Au99bBf0COXEMwqUhIj27RQlH/fJCs42WPPjMcHOOjBx4VYzWG+LZtsZqZoBnh5HS0P8DUk3Npe9wUb6d/kMr3mLlgRcyMyuCmmSZrPkYcEHRD0B/CaapE9qd2gV16xg+CTnr04WvACexwpnEjZJpmJI/SzDYK5g8WyIDdHWtnhTwlW3og0rQGhEKp1I5T+0j8VavmO5S6gfLy9f1a2ownvWBMWEH3ysz/FjpIyhVugvyr+RR5Jyq1nkpwfd75v+A==), and then I look through the AST, just found that if there is a source in export, all `ExportSpeicifer` are `IdentifierName` than `IdentifierReference`, so it cannot reset.

### How to solve?

We should not set current reference flags if the export has `source`, and call `visit_export_specifiers` instead. The code looks like: 
```diff
-        for specifier in &it.specifiers {
-            // `export type { a }` or `export { type a }` -> `a` is a type reference
-            if it.export_kind.is_type() || specifier.export_kind.is_type() {
-                self.current_reference_flags = ReferenceFlags::Type;
-            } else {
-                // If the export specifier is not a explicit type export, we consider it as a potential
-                // type and value reference. If it references to a value in the end, we would delete the
-                // `ReferenceFlags::Type` flag in `fn resolve_references_for_current_scope`.
-                self.current_reference_flags = ReferenceFlags::Read | ReferenceFlags::Type;
-            }
-            self.visit_export_specifier(specifier);
-        }
--
         if let Some(source) = &it.source {
             self.visit_string_literal(source);
+            self.visit_export_specifiers(&it.specifiers);
+        } else {
+            for specifier in &it.specifiers {
+                // `export type { a }` or `export { type a }` -> `a` is a type reference
+                if it.export_kind.is_type() || specifier.export_kind.is_type() {
+                    self.current_reference_flags = ReferenceFlags::Type;
+                } else {
+                    // If the export specifier is not a explicit type export, we conside
r it as a potential
+                    // type and value reference. If it references to a value in the end,
 we would delete the
+                    // `ReferenceFlags::Type` flag in `fn resolve_references_for_current
_scope`.
+                    self.current_reference_flags = ReferenceFlags::Read | ReferenceFlags
::Type;
+                }
+                self.visit_export_specifier(specifier);
+            }
         }
```

Feel free to make it better!

**Comment by camc314** [comment]
ohhh @Dunqing this makes so much sense, thanks for taking the time to help investigate this.
but wow this is complex.

So basically what was happening is:

1. semantic was looping over all export specifiers
2. it was then setting `current_reference_flags`
3. however as these were re-exports (IdentifierName vs IdentifierReference), semantic did not push this in as a reference
4. this meant that stale `current_reference_flags` ended up hanging around for the next time an ident reference was found.

Dunqing fixes this by:
1. only setting `self.current_reference_flags` if `source` is `none` (when source is non, the export specifiers **must** be `IdentifierReference`s over `IdentifierNames` 
2. this means that the value `current_refernce_flags` is taken, replacing it with empty flags.
3. this means that the next identifierReference has the correct flags 🎉

**Comment by Dunqing** [comment]
> ohhh @Dunqing this makes so much sense, thanks for taking the time to help investigate this. but wow this is complex.
> 
Yes, it is complex and unintuitive, but I haven't found a way to make it better, If you are interested in finding a way to sort it out, do it!

**Review by Dunqing — APPROVED** [review]

**Comment by Boshen** [comment]
### Merge activity

* **Dec 27, 9:19 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Dec 27, 9:19 AM EST**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).
